csharp/extractor/Semmle.Extraction.CSharp.DependencyFetching/NugetPackageRestorer.cs

```diff
@@ -231,33 +232,42 @@ private IEnumerable<string> RestoreSolutions(out IEnumerable<string> assets)
         /// <summary>
         /// Executes `dotnet restore` on all projects in projects.
         /// This is done in parallel for performance reasons.
-        /// Populates assets with the relative paths to the assets files generated by the restore.
+        /// Populates dependencies with the relative paths to the assets files generated by the restore.
         /// </summary>
         /// <param name="projects">A list of paths to project files.</param>
-        private void RestoreProjects(IEnumerable<string> projects, out IEnumerable<string> assets)
+        private void RestoreProjects(IEnumerable<string> projects, out List<DependencyContainer> dependencies)
         {
             var successCount = 0;
             var nugetSourceFailures = 0;
-            var assetFiles = new List<string>();
+            List<DependencyContainer> collectedDependencies = [];
             var sync = new object();
-            Parallel.ForEach(projects, new ParallelOptions { MaxDegreeOfParallelism = DependencyManager.Threads }, project =>
+            var projectGroups = projects.GroupBy(Path.GetDirectoryName);
+            Parallel.ForEach(projectGroups, new ParallelOptions { MaxDegreeOfParallelism = DependencyManager.Threads }, projectGroup =>
             {
-                logger.LogInfo($"Restoring project {project}...");
-                var res = dotnet.Restore(new(project, PackageDirectory.DirInfo.FullName, ForceDotnetRefAssemblyFetching: true));
-                lock (sync)
+                var assets = new Assets(logger);
+                foreach (var project in projectGroup)
                 {
-                    if (res.Success)
-                    {
-                        successCount++;
-                    }
-                    if (res.HasNugetPackageSourceError)
+                    logger.LogInfo($"Restoring project {project}...");
+                    var res = dotnet.Restore(new(project, PackageDirectory.DirInfo.FullName, ForceDotnetRefAssemblyFetching: true));
+                    assets.AddDependenciesRange(res.AssetsFilePaths);
+                    lock (sync)
                     {
-                        nugetSourceFailures++;
+                        if (res.Success)
+                        {
+                            successCount++;
+                        }
+                        if (res.HasNugetPackageSourceError)
+                        {
+                            nugetSourceFailures++;
+                        }
                     }
-                    assetFiles.AddRange(res.AssetsFilePaths);
+                }
+                lock (sync)
```

Comment:
We're using the `sync` object for multiple locking purposes. Maybe you could change the `collectedDependencies` to be a `ConcurrentBag<>`.